chain/messagepool/messagepool.go

```diff
@@ -763,7 +763,30 @@ func (mp *MessagePool) Add(ctx context.Context, m *types.SignedMessage) error {
 		<-mp.addSema
 	}()
 
-	mp.curTsLk.Lock()
+	//Ensure block calculation is cached without holding the write lock
+	mp.curTsLk.RLock()
+	tmpCurTs := mp.curTs
+	mp.curTsLk.RUnlock()
+	_, _ = mp.api.GetActorAfter(m.Message.From, tmpCurTs)
+	_, _ = mp.getStateNonce(ctx, m.Message.From, tmpCurTs)
+
+	cacheSecondTime := true
+	//if the newly acquired Ts is not the one we just cached, let go of the lock, cache it and open the lock again and repeat....
+	for cacheSecondTime {
+		mp.curTsLk.Lock()
+		writeCurTs := mp.curTs
+
+		if writeCurTs == tmpCurTs {
+			break // we have this cached we can skip
+		}
+		mp.curTsLk.Unlock()
+		tmpCurTs = writeCurTs
+		_, _ = mp.api.GetActorAfter(m.Message.From, tmpCurTs)
+		_, _ = mp.getStateNonce(ctx, m.Message.From, tmpCurTs)
+		cacheSecondTime = false
```

Comment:
Are we sure we wanna set this to false? Do we not wanna loop again, so as to check whether curTs has changed?

I think we only ever exit this loop at the break on line 780, at which point we can be confident we still hold the curTsLk, but I'm not completely sure.

Reply:
Hey! I rewrote the change and commited it here: 2fa95a09be11deeeadb7fdd5401814178c6d9307 Steb reviewed approves